### ISSUE TITLE
Jupyter Notebook

### ISSUE THREAD

**Issue by blahster** [HubPress/hubpress.io#406]
Hubpress is really nice!
Is there a way to upload a Jupyter Notebook (.ipynb) and have it use the current HubPress theme?
Or embed portions of a Jupyter Notebook in a post?

**Comment by anthonny** [comment]
Hi @blahster 

> Hubpress is really nice!

Thank you!

HubPress do not support Jupyter and to be honest i do not know Jupyter.
I do not know if you can embed a portion of a Jupyter Notebook with javascript ?

**Comment by blahster** [comment]
I see.
I can save a Jupyter Notebook as html, would it be possible to embed that into a post?

**Comment by iacchus** [comment]
Did you managed to do it? It is possible (https://iacchus.github.io/2016/07/27/Testing-Jupyter-Notebook-Embed.html I will delte this later).

1. Go to https://gist.github.com, create something with any name.

2. Clone the repository (all gist are git repositories) clicking in the button Embed / Clone via SSH.

3. Save the .ipynb to that repo, add, commit, and push (delete all directoryes created by Jupyter before, or it won't work.)

4. Login to HubPress, the AsciiDoc command to embed is a html block is this way:

```
++++
<!-- embed code goes here, for example: -->
<script src="https://gist.github.com/iacchus/35fcee9ca4fe705ba2b2f4dd0951e273.js"></script>
++++
```

Publish and that's it. :)

**Comment by blahster** [comment]
@iacchus that worked very well! Thank you.
Any idea how I can increase the width of the article (with the embedded notebook)? It looks squished up right now :)

**Comment by iacchus** [comment]
@blahster The gist will fit to you theme, so the best way would be to change your theme css to enlarge the post container. But this will enlarge your text too, so I think it will be better choose the best size.

I ended up writing a little how to.

Do this, 

1. Go to your blog with the embedded Jupyter and some dummy placeholder text out of the nb (ie., post normal text) with blocks of text large the sufficient for being wrapped by the theme. (This way you can choose the better proportion that is not too large for text an  not too stretched for the jupyter nb.)

    ![](https://i.imgur.com/O8D0ABB.png)

2. Open your browser's code inspector (Ctrl+Shift+I in chromium/chrome).

    ![](https://i.imgur.com/HHpAcy7.png)

3. And here you should find the css classes in your theme that are setting the size of your post container. I'll tell you what you can do in Chromium/Chrome but in other browsers should be something similar. 

    **In Code Inspector, when you point the mouse to a line of an element, your see certain colors in the browser's viewport.**

    **Blue means the area inside of the highlighted element. 
Red means margin/space outside the element. (this can be set by a css class in it.)
Green means padding, area inside the element but not used by its content.**

    Here, I put the mouse pointer inside the element `<main id="content" class="content" role="main">` which shows a 100% blue width, meaning that that is not this element which is constricting our text width.

    ![](https://i.imgur.com/ELeYzSB.png)

    Then, in the element tree showing in the inspector I point the mouse the `<main>`'s first child, `<article id="" class="post first">` in my theme, which shows this:

    ![](https://i.imgur.com/aS3Kaf5.png)

    The **red is that element's margin, which means that this is probably the element which has a css class setting our post text/gist width, that we must tinker with the values**.

4. Click that element so select it.  Then we go to the left side to the inspector, in the styles tab and try to find out which classes may me setting that element `width` and `max-width` values. Here we can see that it is .post class.

    ![](https://i.imgur.com/gNQP8cL.png)

    Here we see three things, 1. the filename of the theme css which will need to be changed, which here is `screen.css`, 2. the name of the class which will need to be changed, which here is `.post` and the values `width` and `max-width`. (`max-width` overwrites `width`)

5. Click that values and change them till you find which one fits better.

    800px (default was 700px)
    ![](https://i.imgur.com/Wc84WS0.png)

    900px
    ![](https://i.imgur.com/AosKlF9.png)

6. Then after you decide go to your HupPress repo, enter `/themes/YOUR_CURRENT_THEME/assets/css/` and there you open the file you see in point 4, for the example here was `screen.css`. Save, commit, push, and reload you browser. (css changes don't need to rebuild theme template)

Then would be good if you checked the template's responsiveness in medium/small viewports (tablet/phone) to see if you didn't broke anything. If this happens, search that css (`screen.css` for this example) file for `@media` string and inside that specific @media change the class (`.post` in this example) accordingly.

**Comment by blahster** [comment]
@iacchus, thank you for the detailed steps. It worked perfectly.
You're right, 800px looks much better; I'm using roon as the theme, so edited max-width for `article`.

I went through your EDIT2 but couldn't figure out how to call the `onload` for the iframe to show the full height. But thank you for trying.

@anthonny, please see if you can make notebooks first class citizens in HubPress (love the product!)

**Comment by iacchus** [comment]
@blashter I don't know if this is possible. A person who posted a comment on first link said:

> This will not work if the iframe contains content from another domain because of the [Same Origin Policy](https://en.wikipedia.org/wiki/Same-origin_policy), there is another Question on SO with some cross domain solutions

Javascript is needed so that the parent html can take child's html size to resize the iframe. The policy which that guy is saying is a security measure that browsers implement, so that an attacker cannot load a site full height full width in a iframe, and impersonate another site so that by using javascript he can steal data from the fake upper site of the legitimate one running inside the iframe. I don't know if there is a mean they can setup the webserver to override this policy. (as `USERNAME.github.io` and `gist.github.com` are different domains but both are from the same owner.) 

But I am trying.. otherwise I can think of the notebook saved as html and embedded directly (the file being inside the HubPress repo much like `images/`.)

**Comment by iacchus** [comment]
I think I could embed it directly using [rawgit.com](https://rawgit.com), and also to inject my theme's css in it. Example: (I will delete this later: https://iacchus.github.io/2016/08/03/More-Stuff-Implementing-html-ipython-notebook-in-Hub-Press.html) this way

1. In hubpress repo root, created a directory `ipynb-html`.

2. Used this script (highlighted, lines 25 to 64 in my theme). This should be put anywhere in document.load(), here using jQuery for this:
https://github.com/iacchus/ghostdance/blob/30fde23335486a2b2187b417c71f1c4ada6f7982/assets/js/index.js#L25-L64
Maybe somebody using this will have to change lines 44 and 47 to fit it's theme's css. (file at line 47 see point 5.)

3. Put the file inside the directory created in 1., for example `Untitled3.html` so that i had 'ipynb-html/Untitled3.html'.

4. Post an `iframe` with the class `ipynb-embed` and with attribute `data-filename` with the name of the file (omitting the directory name), very simple:
    ```
++++
<iframe class="ipynb-embed" data-filename="Untitled3.html"></iframe>
++++
```

5. I created this css (inside theme's `assets/css/`) to override some bugs with notebook on mobile and in my theme:
https://github.com/iacchus/ghostdance/blob/30fde23335486a2b2187b417c71f1c4ada6f7982/assets/css/ipynb-custom.css

I implemented this to my theme, I just would like to find a better way to embed, for example:

`ipynb::Untitled3.html[]`
